crates/mdk-core/src/messages.rs

```diff
@@ -2743,4 +2750,1087 @@ mod tests {
+        let (creator, members, admins) = create_test_group_members();
+        let group_id = create_test_group(&mdk, &creator, &members, &admins);
+
+        // Create a message
+        let rumor = create_test_rumor(&creator, "Test message with relay issues");
+        let message_event = mdk
+            .create_message(&group_id, rumor)
+            .expect("Failed to create message");
+
+        // Process the message
+        // Note: In a real scenario with actual relay connections, we would simulate
+        // relay failures. For this unit test, we verify that message processing
+        // itself is resilient and doesn't depend on relay availability for processing.
+        let result = mdk.process_message(&message_event);
+        assert!(
+            result.is_ok(),
+            "Message processing should succeed regardless of relay state"
+        );
+
+        // Verify the message was stored locally
+        let messages = mdk.get_messages(&group_id).expect("Failed to get messages");
+        assert_eq!(messages.len(), 1, "Message should be stored locally");
+
+        // Verify message content is intact
+        let stored_message = &messages[0];
+        assert_eq!(
+            stored_message.content, "Test message with relay issues",
+            "Message content should be preserved"
+        );
+    }
+
+    /// Test message deduplication across multiple relay sources
+    ///
+    /// Validates that messages with the same ID from different relays are
+    /// deduplicated correctly.
+    #[test]
+    fn test_message_deduplication_across_relays() {
+        let mdk = create_test_mdk();
+        let (creator, members, admins) = create_test_group_members();
+        let group_id = create_test_group(&mdk, &creator, &members, &admins);
+
+        // Create a message
+        let mut rumor = create_test_rumor(&creator, "Deduplicated message");
+        let rumor_id = rumor.id();
+        let message_event = mdk
+            .create_message(&group_id, rumor)
+            .expect("Failed to create message");
+
+        // Process the message multiple times (simulating multiple relay sources)
+        for i in 0..5 {
+            let result = mdk.process_message(&message_event);
+            // Each processing should either succeed (idempotent) or fail gracefully
+            if i == 0 {
+                assert!(result.is_ok(), "First processing should succeed");
+            }
+            // Subsequent processings may succeed or fail, but shouldn't panic
+        }
+
+        // Verify we only have one message stored
+        let messages = mdk.get_messages(&group_id).expect("Failed to get messages");
+        assert_eq!(
+            messages.len(),
+            1,
+            "Should have exactly 1 message despite multiple processing attempts"
+        );
+
+        // Verify the message ID matches
+        assert_eq!(
+            messages[0].id, rumor_id,
+            "Stored message should have correct ID"
+        );
+    }
```

Comment:
Isn't this basically doing the same as the `test_duplicate_message_from_multiple_relays` test?